File: algorithms/median_insert.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

int list[5];
int median;
int median_index;


int print_list() {
	int i;
	for (i = 0; i < sizeof(list); i ++) {
		printf("%d ", list[i]);
	}
	printf("\n");
}


int get_list() {
	printf("Enter 5 integers:\n");
	int num_inputs = scanf("%d %d %d %d %d", list, &list[1], &list[2], &list[3], &list[4]); 
	if ( num_inputs == 5) return 0;
	else return -1;
}



// NOTE: to insert, the list size should be incremented
//	unfortunately, the list variable is a 5-integer variable
//	and cannot be changed
// NOTE: hence we need to create a new larger list for every new input
void insert() {
	// NOTE: read new input
	int new_input;
	printf("Enter a new integer:\n");
	scanf("%d", &new_input);
		

	// NOTE: attempt 1, local variable list
	int new_list[sizeof(list) + 1];
	// copy all present elements
	int i;
	for (i = 0; i < sizeof(list); i ++) {
		new_list[i] = list[i];
	}
	// now add new input to end of list
	// note that the value of i = index of last element
	new_list[i] = new_input;
	// finally assign list to point to this new list
	// NOTE: compilation will fail here; we cannot do this
	//	because list is already a fixed symbol in this binary
	//	that is known to reference 5 consecutive integers
	list = new_list;
	// print to confirm
	print_list();
}


int my_sort_list() {
	int i, j, x, y;
	for (i = 1; i < sizeof(list); i ++) {
		x = list[i];
		for (j = i - 1; j >= 0; j--) {
			y = list[j];
			if (x < y) {
				list[j] = x;
				list[j+1] = y;
			}
			print_list();
		}
	}

	return 0;
}


void get_median() {
	median_index = (sizeof(list)/sizeof(int))/2;
	median = list[median_index];
}


int main() {
	get_list();
	printf("Successfully read integer list: ");
	print_list();

	my_sort_list();
	printf("Successfully sorted list: "),
	print_list();

	insert();
	printf("Successfully inserted to list: "),
	print_list();

	get_median();
	printf("\n Median: %d\n", median);
	

	return 0;
}

File: algorithms/median_insert.py
# NOTE: everything is easier in python for one main reason
#	--> every variable in python is a dynamic one
#	and can be reassigned dynamically
# NOTE: in python, a variable is just a name/id
#	--> it is not tied to a data type or even a memory location

# NOTE: this statement creates a dynamic list object
#	--> this is identical to creating a pointer to data of a certain
#	type in C
int_list = []

# new list_size
list_size = 0

median = 0
median_index = 0



# NOTE: the fact that all variables in python are dynamically allocated
#	is a reason why we have to declare global scope
#	if we want to use and update variables globally
def get_list():
	global int_list
	global list_size

	print("Enter 5 integers: ")
	str_list = input().split()
	# new list_size
	list_size = len(str_list)
	for i in range(0, list_size):
		# NOTE: append() function acts like malloc in C
		#	--> it dynamically allocates a new integer
		#	variable and adds it to the chunk of memory
		#	pointed to by int_list
		# NOTE: append() does malloc, then casts to data type,
		#	then updates the value of allocated memory
		#	to the parameter value
		int_list.append(int(str_list[i]))


def print_list():
	global int_list
	print(int_list)


def sort_list():
	global int_list
	global list_size

	print("Sorting input list ...")
	# new list_size
	for i in range(1, list_size):
		x = int_list[i]
		for j in range(i - 1, -1, -1):
			y = int_list[j]
			if (x < y):
				int_list[j] = x
				int_list[j + 1] = y
			print_list()


def insert():
	global int_list
	global list_size

	print("Insert a new integer: ")
	new_entry = int(input())

	# NOTE: can append without explicitely allocating a larger
	#	array; the implementation of append() function 
	#	already does that
	int_list.append(new_entry)
	list_size = list_size + 1

	# NOTE: sort list will fail without new list size
	sort_list()
	

def get_median():
	global median
	global median_index
	global list_size

	median_index = int(list_size/2)
	median = int_list[median_index]


def main():
	get_list()
	print("Input list = ")
	print_list()

	sort_list()
	print("Sorted list = ")
	print_list()

	insert()
	print("New list  = ")
	print_list()

	get_median()
	print("Median = ", median)


main()



